ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/HardFork/Combinator/Ledger.hs

```diff
@@ -755,3 +825,423 @@ injectLedgerEvent index =
+        , HasHardForkTxOut xs
+        )
+  => Index xs x
+  -> LedgerTables (LedgerState                x  ) mk
+  -> LedgerTables (LedgerState (HardForkBlock xs)) mk
+injectLedgerTables idx =
+    LedgerTables
+  . mapKeysMK injTxIn
+  . mapMK injTxOut
+  . getLedgerTables
+  where
+    injTxIn :: Key (LedgerState x) -> Key (LedgerState (HardForkBlock xs))
+    injTxIn = injectCanonicalTxIn idx
+
+    injTxOut :: Value (LedgerState x) -> Value (LedgerState (HardForkBlock xs))
+    injTxOut = injectHardForkTxOut idx
+
+distribLedgerTables ::
+     forall xs x mk. (
+          CanMapKeysMK mk
+        , Ord (Key (LedgerState x))
+        , HasCanonicalTxIn xs
+        , CanMapMK mk
+        , HasHardForkTxOut xs
+        )
+  => Index xs x
+  -> LedgerTables (LedgerState (HardForkBlock xs)) mk
+  -> LedgerTables (LedgerState                x  ) mk
+distribLedgerTables idx =
+    LedgerTables
+  . mapKeysMK (distribCanonicalTxIn idx)
+  . mapMK (distribHardForkTxOut idx)
+  . getLedgerTables
+
+{-------------------------------------------------------------------------------
+  HardForkTxIn
+-------------------------------------------------------------------------------}
+
+-- | Defaults to a 'CannonicalTxIn' type, but this will probably change in the
+-- future to @NS 'WrapTxIn' xs@. See 'HasCanonicalTxIn'.
+type instance Key   (LedgerState (HardForkBlock xs)) = CanonicalTxIn xs
+
+-- | Canonical TxIn
+--
+-- The Ledger and Consensus team discussed the fact that we need to be able to
+-- reach the TxIn key for an entry from any era, regardless of the era in which
+-- it was created, therefore we need to have a "canonical" serialization that
+-- doesn't change between eras. For now we are requiring that a 'HardForkBlock'
+-- has only one associated 'TxIn' type as a stop-gap, but Ledger will provide a
+-- serialization function into something more efficient.
+type HasCanonicalTxIn :: [Type] -> Constraint
+class ( Show (CanonicalTxIn xs)
+      , Ord (CanonicalTxIn xs)
+      , NoThunks (CanonicalTxIn xs)
+      ) => HasCanonicalTxIn xs where
+  data family CanonicalTxIn (xs :: [Type]) :: Type
+
+  -- | Inject an era-specific 'TxIn' into a 'TxIn' for a 'HardForkBlock'.
+  injectCanonicalTxIn ::
+    Index xs x ->
+    Key (LedgerState x) ->
+    CanonicalTxIn xs
+
+  -- | Distribute a 'TxIn' for a 'HardForkBlock' to an era-specific 'TxIn'.
+  distribCanonicalTxIn ::
+    Index xs x ->
+    CanonicalTxIn xs ->
+    Key (LedgerState x)
+
+  encodeCanonicalTxIn :: CanonicalTxIn xs -> CBOR.Encoding
+
+  decodeCanonicalTxIn :: forall s. CBOR.Decoder s (CanonicalTxIn xs)
+
+{-------------------------------------------------------------------------------
+  HardForkTxOut
+-------------------------------------------------------------------------------}
+
+-- | Defaults to the 'HardForkTxOut' type
+type instance Value (LedgerState (HardForkBlock xs)) = HardForkTxOut xs
+
+-- | This choice for 'HardForkTxOut' imposes some complications on the code.
```

Comment:
Compiling them now

Reply:
That took more effort than expected, because the haddocks.sh script doesn't work for windows (yet)

Reply:
![image](https://github.com/user-attachments/assets/0a61ac14-190b-4bed-9fc5-2af50851d154)
Pasting a screenshot because they are SVGs and Github doesn't render those by default.